crates/plugins/src/manager.rs

```diff
@@ -0,0 +1,219 @@
+use crate::{
+    lookup::PluginLookup,
+    manifest::{check_supported_version, PluginManifest, PluginPackage},
+    store::PluginStore,
+    SPIN_INTERNAL_COMMANDS,
+};
+
+use anyhow::{anyhow, bail, Context, Result};
+use sha2::{Digest, Sha256};
+use std::{
+    fs::{self, File},
+    io::{copy, Cursor},
+    path::{Path, PathBuf},
+};
+use tempfile::{tempdir, TempDir};
+use url::Url;
+
+// Url scheme prefix of a plugin that is installed from a local source
+const URL_FILE_SCHEME: &str = "file";
+
+/// Location of manifest of the plugin to be installed.
+pub enum ManifestLocation {
+    /// Plugin manifest can be copied from a local path.
+    Local(PathBuf),
+    /// Plugin manifest should be pulled from a specific address.
+    Remote(Url),
+    /// Plugin manifest lives in the centralized plugins repository
+    PluginsRepository(PluginLookup),
+}
+
+/// Provides accesses to functionality to inspect and manage the installation of plugins.
+pub struct PluginManager {
+    store: PluginStore,
+}
+
+impl PluginManager {
+    /// Creates a `PluginManager` with the default install location.
+    pub fn default() -> anyhow::Result<Self> {
+        let store = PluginStore::default()?;
+        Ok(Self { store })
+    }
+
+    /// Returns the underlying store object
+    pub fn store(&self) -> &PluginStore {
+        &self.store
+    }
+
+    /// Installs the Spin plugin with the given manifest If installing a plugin from the centralized
+    /// Spin plugins repository, it fetches the latest contents of the repository and searches for
+    /// the appropriately named and versioned plugin manifest. Parses the plugin manifest to get the
+    /// appropriate source for the machine OS and architecture. Verifies the checksum of the source,
+    /// unpacks and installs it into the plugins directory.
+    /// Returns name of plugin that was successfully installed.
+    pub async fn install(
+        &self,
+        plugin_manifest: &PluginManifest,
+        plugin_package: &PluginPackage,
+    ) -> Result<String> {
+        let target = plugin_package.url.to_owned();
+        let target_url = Url::parse(&target)?;
+        let temp_dir = tempdir()?;
+        let plugin_tarball_path = match target_url.scheme() {
+            URL_FILE_SCHEME => PathBuf::from(target_url.path()),
```

Comment:
Something like this might behave better on Windows:

```suggestion
            URL_FILE_SCHEME => target_url.to_file_path()
                .map_err(|_| anyhow!("Invalid file URL: {target_url:?}")),
```

https://docs.rs/url/2.3.1/url/struct.Url.html#method.to_file_path

Note the oddity here that `to_file_path` returns `Result<PathBuf, ()>` and since `()` doesn't `impl Error` it means `context` doesn't work... :shrug: